subsys/bluetooth/services/gatts_client.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <sys/byteorder.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/gatts_client.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gatts_c, CONFIG_BT_GATTS_CLIENT_LOG_LEVEL);
+
+enum {
+	GATTS_SC_INDICATE_ENABLED
+};
+
+int bt_gatts_client_init(struct bt_gatts_client *gatts_c)
+{
+	if (!gatts_c) {
+		return -EINVAL;
+	}
+
+	memset(gatts_c, 0, sizeof(struct bt_gatts_client));
+
+	return 0;
+}
+
+static void gatts_reinit(struct bt_gatts_client *gatts_c)
+{
+	gatts_c->conn = NULL;
+	gatts_c->handle_sc = 0;
+	gatts_c->handle_sc_ccc = 0;
+	gatts_c->state = ATOMIC_INIT(0);
+}
+
+int bt_gatts_handles_assign(struct bt_gatt_dm *dm,
+			    struct bt_gatts_client *gatts_c)
+{
+	const struct bt_gatt_dm_attr *gatt_service_attr =
+		bt_gatt_dm_service_get(dm);
+	const struct bt_gatt_service_val *gatt_service =
+		bt_gatt_dm_attr_service_val(gatt_service_attr);
+	const struct bt_gatt_dm_attr *gatt_chrc;
+	const struct bt_gatt_dm_attr *gatt_desc;
+
+	if (bt_uuid_cmp(gatt_service->uuid, BT_UUID_GATT)) {
+		return -ENOTSUP;
+	}
+
+	LOG_DBG("GATT Service found");
+
+	gatts_reinit(gatts_c);
+
+	gatt_chrc = bt_gatt_dm_char_by_uuid(dm, BT_UUID_GATT_SC);
+	if (!gatt_chrc) {
+		return -EINVAL;
+	}
+
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc,
+					    BT_UUID_GATT_SC);
+	if (!gatt_desc) {
+		return -EINVAL;
+	}
+	gatts_c->handle_sc = gatt_desc->handle;
+
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc, BT_UUID_GATT_CCC);
+	if (!gatt_desc) {
+		return -EINVAL;
+	}
+	gatts_c->handle_sc_ccc = gatt_desc->handle;
+
+	LOG_DBG("Service Changed characteristic found.");
+
+	/* Finally - save connection object */
+	gatts_c->conn = bt_gatt_dm_conn_get(dm);
+
+	return 0;
+}
+
+static uint8_t bt_gatts_sc_indicate_callback(struct bt_conn *conn,
+					     struct bt_gatt_subscribe_params *params,
+					     const void *data, uint16_t length)
+{
+	struct bt_gatts_client *gatts_c;
+	struct bt_gatts_handle_range handle_range;
+	int err;
+	const uint8_t *handles_data = (const uint8_t *)data;
+
+	/* Retrieve GATT Service client module context. */
+	gatts_c = CONTAINER_OF(params, struct bt_gatts_client, indicate_params);
+
+	if (gatts_c->indicate_cb) {
+		err = (length == 4) ? 0 : -EINVAL;
```

Comment:
Please add define for length instead of magic number

Reply:
Will do. The macro will be BT_GATTS_SERVICE_CHANGED_VALUE_LENGTH.